### ISSUE TITLE
Doesn't build with devkitARM r46, libnds 1.6.2 and maxmod 1.0.10

### ISSUE THREAD

**Issue by WinterMute** [ApacheThunder/NTR_Launcher_Old#5]
Please make more effort to maintain compatibility with upstream rather than using bodged up forks.

This creates major support headaches and wastes time and effort that could better be spent improving the tools.

**Comment by ApacheThunder** [comment]
I'm aware latest devkitarm. However my build environment is setup to build Ahezard's other projects.

So yeah....

I'm waiting on him to port nds-bootstrap over to latest devkitarm. Until that happens, I can't update to latest devkitarm as at that point I would no longer be able to keep my stuff maintained at the same time as being able to compile his stuff.

So yeah I can see why it would be better to have it on latest devkitarm and not on custom/modified libraries. As I recall the only stuff NTR_Launcher depends on ahezard's libnds fork for is pre-defined slot reset code and pre-defined SCFG register code. I haven't worked on this project in awhile (with only a small addition I recently added).

I could probably separate my project's reliance on his fork relatively easily, however actually making it compile correctly with latest devkitarm may not be within my abilities....Though I haven't actually tried yet, so not sure. Waiting on Ahezard to move bootstrap over to r46, so won't know until that happens. :P

I have not used latest devkitarm at all yet. I may need to enlist the help of someone else to port it over to latest devkitarm. My coding knowledge is very limited and still is. Frankly surprised I managed to get this far with this. :P

NTR Launcher was a redo of a separate "launcher" I found on filetrip. A long time back someone made a stripped down slot-1 launcher from NitroHax and didn't even release the source code. This had me miffed a little as I wanted to improve it. I ended up re-doing what he did as I had a basic idea on how it was done.

I may retire the main build in favor of the alt-build when I do port this over eventually. The alt-build is able to boot in both NTR and TWL mode correctly while the main build is not able to handle TWL mode. I haven't heard any compatibility issues TWL version has that are different from main build, so I don't really see a reason to maintain both anymore....

**Comment by ahezard** [comment]
Please calm down a little bit, you are not helping. This will be done but considering you are doing only a release of linbnds once a year I do not feel an urgence to do it. We are all doing this on free time anyway.

**Comment by ApacheThunder** [comment]
No need for the hostilities. I welcome WinterMute's input on this. :P

I actually do want to use r46 at some point. It has a lot of new stuff I'd like to make use of anyways. I always felt the current way my project compiles is a bit hacky. Especially the way arm7 code/ram entry address stuff is handled. :P

Anyways as of this commit:

https://github.com/ApacheThunder/NTR_Launcher/commit/100866387491bb41c90700327522a8f3eded42bb

The "alt build" branch of NTR Launcher no longer requires Ahezard's fork of of libnds. The only function I had to give up as a result was his force DS mode touch screen function which I don't think it needs anyways. NTR Launcher always expects to boot in TWL mode. It switches to NTR mode prior to launching the game/cart. (though TWL mode remaining in place has been a feature of the alt branch and is available via a setting in the ini file. It breaks some games and causes audio/touch screen issues for all others, but it's there when this eventually gets support for DSi enhanced games as remaining in TWL mode will be important for those)

I will retire the main branch however. It does not correctly operate in NTR mode if booted in NTR mode from a separate application due to unlocked SCFG on arm7 if said application had enabled unlocked arm7 SCFG in NTR mode. This results in libnds code setting up DSi framebuffer/touchscreen code despite NTR mode being active. Maybe this was fixed in r46? Doesn't matter though since main build doesn't work in TWL mode which is important for future TWL cart support. So there's no reason to continue maintaining the main build when I see no disadvantage to the alt build.

This has no practical impact on my project especially since this issue would in theory not crop up if booted in NTR mode from a flashcart on a DSi...(though it would be useless in this circumstance) and it's not currently possible to run with unlocked DSi hardware in NTR mode on a DSi anyways so this issue wouldn't apply there. Not to mention this has no practical use on an original DS since hotswapping carts has always been possible with that console.

I do recall there are forks a few people made where they use my launcher as a forwarder for flashcarts. In those instances that issue would crop up if they launch my launcher indirectly from something that switches to NTR mode prior to booting NTR Launcher.  In those instances I would just suggest they get their projects to remain in TWL mode when booting NTR Launcher. NTR Launcher should ultimately make the switch anyways. Since DSi hardware doesn't actually have to be disabled until just before it boots the game which mine already does. :D

Main build has no support for TWL mode either so running DSi Enhanced carts in TWL mode would be out of the question as well.

As it stands now it will be easier to port over to latest devkitarm libraries. however I'm not sure I can do that myself. If someone makes a push request for it, I'll setup a new build environment on my laptop and build it there so I don't have to worry about conflicting with my main build environment I use for Ahezard's stuff as I won't be able to build his stuff anymore once I move to it.

I'm not sure how I'm going to deal with the project not being forked properly from official NitroHax's github source. Unless I can do so without deleting the existing project from Github then I guess that means I no plans on remedying this. There are 4+ forks of NTR Launcher floating around and if I nuke this project to properly rebase from NitroHax as a proper fork then that may disrupt their forks. That and forks can't have "issue" threads and such opened. Seems Github limits a few things with forks in respect to the original. So if I had forked it from official Nitrohax (which wasn't possible at the time since he didn't have it on his github when I created this project) then you wouldn't have been able to open these issues threads and let me know of the issues you have mentioned. (though only specifically applies to the other issue you opened and not this one. That other issue you opened would not have needed to be opened if it was forked properly anyways.  :P )

So in that respect I probably won't correct this. I have made it clear what code NTR Launcher was based from though.

EDIT: Just noticed maxmod was updated? lol well at least that shouldn't be too hard to fix. I really only use that for the boot chime for the DS/DSi boot animation this project uses. :P

**Comment by WinterMute** [comment]
Something to consider here is that if you have unlocked arm7 scfg then you're not in NTR mode, you're still in DSi mode.

libnds has no "DSi framebuffer" and it hardly seems unreasonable to use TWL mode touchscreen code while in TWL mode. I'm not sure I see the relevance here though, isn't this purely about booting slot1 cards?

Using this code as a "forwarder" for pirate carts seems rather silly. Given some relatively minor patching NitroHax could do that just fine without requiring a bunch of special builds for various cards. If it doesn't already "autoboot" then there should be an option added to do just that.

The logo stuff you have in here could be done so much better as well. You don't need that many images just to do a fade. I'll see if I can dig out some code to show you what I mean.

Ultimately the community would be better served by improving NitroHax proper rather than ripping things out of it. In theory it should be possible to boot & patch carts in TWL mode with a little bit of work too.

**Comment by ApacheThunder** [comment]
Well yes I suppose "technically" it's not true NTR mode if I leave TWL hardware active. But we're arguing over technicalities at this point. By default NTR Launcher does lock SCFG_EXT on arm7 and arm9 using default values expected for NTR mode. This results in the same NTR environment that Nintendo intended on their 3DS platform for DS game backwards compatibility.

The default INI file I provide with this has the "ENABLESD" option disabled. When that option is off, SCFG is locked on both arm7 and arm9 and uses standard NTR values.

As for the touchscreen issues now that I think aback on last time I messed around with it. TWL mode alone doesn't really break touchscreen for NTR games (it does however cause audio problems in most games. Games don't completely break unless you enable DSi ram space since NTR games expect specific addresses to be mirrors of the 4mb ram space. Changing this obviously breaks things) 

That only happens if DSi touchscreen is enabled as a result of certain settings being used in the SCFG_EXT section of the DSi Extended Header of the DSiWare SRL being used to boot said game/launcher. How DSi Touchscreen stuff is triggered and how the NTR version of that is triggered isn't well documented at all, hence I use SCFG_EXT values that don't enable it.

I noticed your new r46 of devkitarm/libnds compiles SRLs that have the SRL use DSi Enhanced mode instead of DSi Exclusive mode (as set in offset 0x12 in the NTR header) . Or at least that's the mode your bootstrap CIA is using, so this in turn impacts anything you would boot from it. 

I wasn't aware of any functional difference. I can't recall if even the SCFG_ROM register even looks any different from the two modes. Maybe you did this for compatibility with homebrew running from slot-1. But it's kinda wierd to see DSiWare use that mode. I've never seen any official DSiWare SRLs use that mode so....

You also end up triggering DSi touchscreen as result of the SCFG_EXT values you set, so for the moment it's not possible to switch back. This would break older homebrew compiled on older libraries. Though I suppose you don't really have to provide compatibility for old projects though. :P

As for the "framebuffer" issues. I don't really understand the cause of it very well, so I just called it a framebuffer issue because that's what it looked like. If SCFG_EXT is left unlocked, older homebrew misdetects that it's in full TWL mode and uses TWL bios functions for certain stuff I would assume. For example if I launch NTR Launcher in NTR mode from another application that left SCFG unlocked, the boot animation does not appear correctly (top screen remains black while bottom screen remains white though I think the boot chime still worked) UI bugs in other apps are seen as well. hbmenu fails to display the top screen graphics (though still shows text on the top screen correctly though). So obviously I recognize SCFG_EXT being unlocked in NTR mode is non standard and usually I leave it locked by default.

The reason I provide an option to unlock it is that it may be necessary to access 3DS SD from homebrew running in NTR mode on a slot-1 device. SCFG must be unlocked for this (though there are some exceptions I noticed. Like the modified build of WoodR4 Ahezard made could still access SD on locked SCFG if the values used for arm7 had left nand access enabled prior to locking it)

All that aside, updates to libnds/devkitarm are very few and far between. So Ahezard had to make his own changes libnds to do the stuff he wanted to do. He didn't want to wait for libnds to get updated. If he did, he wouldn't be as far along with his projects as well. Perhaps he could have contacted you earlier and arranged to have that sorted out sooner. Not sure. I suppose I'll leave that for you two to sort out.

As for the logo code I'm using.....Yeah I'm aware of how terrible it is. I looked around for example code from other projects, but couldn't really find anything. The closest thing I found was some app that only ran a anmiation and sound. But was coded in PALib....so yeah not touching that with a 9 foot pole...Not that I could, it was never open source anyways.

Lacking any suitable examples, I did it the only way I knew how to given the limited coding knowledge I have. It more or less works. It just takes up a lot more space then it should and is a mess coding wise. But it works so that's what I use for now.  If you manage to dig up something that can help me in that regard, then by all means. I want to fix this mess of a logo code. It takes forever to compile like it is right now anyways. :P

As for the issue with not being forked with NitroHax......Well I suppose I could just delete this and redo it. But that shows lack of consideration to the people who forked my project. Sure the only notable fork I am concerned this will break is a "forwarder" for flashcarts. But understand there are people out there that want to have icons for their favorite DS games on the 3DS Home Menu. Personally I find this kinda silly (especially now that nds-bootstrap/TWLoader is a thing and soon most games will be fully playable off SD with it's own GUI based loader so launching individual game icons from Home Menu seems...superfluous. :P

But I'm not going to nuke their github forks because I don't agree with how their using my project which is already based on someone elses project to begin with. I'm too laid back for that I guess. :P

NTR Launcher was initially made to enable launching really really old flashcarts. But with the stuff I've added later on, it has seen more use in "overclocking" NTR mode by enabling DSi clock speeds on arm9 for NTR games. This more or less works fine for most games and boosts performance of some games like Trackmania games. I see noticably better frame rates with those games with the boosted arm9 clock speed/vram speeds being enabled.

It is really non standard though so I recently noticed a few issues crop up from doing that. Mario 64 for example breaks if sleep mode is triggered while TWL clock speeds are being used.

But as I've mentioned before, default INI file settings for this app has standard NTR speeds/SCFG values from the gitgo, It's up to the end user to enable those options.

Maybe once this project is moved over to libnds 1.6.2/devkitarm r46 I'll refork this properly. But at the moment I'm just not motivated to do that and deal with the issues with the folks who did fork my project already. Maybe I can just warn them about it so as to make sure they have backups then just do it.

I suppose I could have just made NitroHax auto boot and be able to act as a simple cart launcher, but I wanted something that was streamlined for that exact purpose. NitroHax is useful as a homebrew action replay alternative. I did't want to bugger that up with my own stuff so I've kept NTR Launcher a separate thing. I have no plans on changing going back so....

I'll get to having this using latest official devkitarm/libnds when I find the time/help needed to do it. But for now I'm not going to really do anything about the not being forked from NitroHax issue.

**Comment by WinterMute** [comment]
>As for the "framebuffer" issues. I don't really understand the cause of it very well, so I just called it a framebuffer issue because that's what it looked like. If SCFG_EXT is left unlocked, older homebrew misdetects that it's in full TWL mode and uses TWL bios functions for certain stuff I would assume. For example if I launch NTR Launcher in NTR mode from another application that left SCFG unlocked, the boot animation does not appear correctly (top screen remains black while bottom screen remains white though I think the boot chime still worked) UI bugs in other apps are seen as well. hbmenu fails to display the top screen graphics (though still shows text on the top screen correctly though). So obviously I recognize SCFG_EXT being unlocked in NTR mode is non standard and usually I leave it locked by default.

This is just a mismatch with the paged in bios. I'm also not 100% convinced by the current DSi detection code.

https://github.com/devkitPro/libnds/blob/d0eb3b22a4f26dcdcc5b163108dc60aafe3944d7/source/common/syscalls.c
https://github.com/devkitPro/libnds/blob/master/source/common/biosCalls.s#L117-L127


>All that aside, updates to libnds/devkitarm are very few and far between. So Ahezard had to make his own changes libnds to do the stuff he wanted to do. He didn't want to wait for libnds to get updated. If he did, he wouldn't be as far along with his projects as well. Perhaps he could have contacted you earlier and arranged to have that sorted out sooner. Not sure. I suppose I'll leave that for you two to sort out.

Updates to libnds/devkitARM happen as and when they're needed & it makes sense to roll new releases. They tend to get delayed a *lot* by time being wasted sorting out the mess people make when they decide to modify the tools and libraries for their own purposes without even bothering to consult upstream.

>I noticed your new r46 of devkitarm/libnds compiles SRLs that have the SRL use DSi Enhanced mode instead of DSi Exclusive mode (as set in offset 0x12 in the NTR header) . Or at least that's the mode your bootstrap CIA is using, so this in turn impacts anything you would boot from it.

That's basically just a flag , it just lets the loader know that the nds file contains extra binaries which should be loaded in DSi mode. It's not DSi Enhanced/Exclusive Mode, it's Hybrid DS/DSi and DSi exclusive binary.
 
Deleting your project on github will have no effect on anything forked from it. Whatever anyone else has done on their own forks is *entirely* independent of yours. You can't "nuke" anyone else's fork, only your own.

**Comment by ApacheThunder** [comment]
Ahh I understand, so forks won't be harmed from this I guess. I'll remove the old repo shortly then. Not sure where I got the idea it would impact other forks from. :P

**Comment by ApacheThunder** [comment]
Ok I've gotten the build environment setup on my old laptop. The examples provided with devkitpro compile fine and have valid DSi Extended Headers.

So I've attempted to compile NTR_Launcher to get a look at what I'll need to fix. First thing I had to do was remove the bios.s file stuff from arm7/arm9.

But it still hangs up on this error:

linking NTR_Launcher.arm7.elf
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0xd0): undefined reference to `__arm7_lma__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0xe0): undefined reference to `__arm7_start__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0xe4): undefined reference to `__arm7_end__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0xf8): undefined reference to `__arm7i_start__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0xfc): undefined reference to `__arm7i_end__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0x100): undefined reference to `__twl_bss_start__'
c:/devkitpro/devkitarm/bin/../lib/gcc/arm-none-eabi/6.3.0/../../../../arm-none-e
abi/lib/ds_arm7_crt0.o: In function `arm7lma':
(.crt0+0x104): undefined reference to `__twl_bss_end__'
collect2.exe: error: ld returned 1 exit status
make[2]: *** [/d/NDS/NTR_Launcher/arm7/NTR_Launcher.arm7.elf] Error 1
make[1]: *** [build] Error 2
make[1]: Leaving directory `/d/NDS/NTR_Launcher/arm7'
make: *** [arm7/NTR_Launcher.elf] Error 2

I suppose perhaps the other changes shutterbug did isn't working here. I also get a bunch of warning about the REG_SCFG_MC and other stuff already being defined in libnds. Cool looks like that's officially in libnds now so I'll just remove all those defines I don't need anymore. :D

I'll look at the makefile to hbmenu next time I try to compile this and restore original references to the stock ld/specs files as well. Pending those two issues I'm ready to get this working on clean r46 devkitarm/libnds. Just have to resolve the compile issues and test that this actually works properly with the way new libnds restructures things. You said NitroHax still worked so that means this should too. :D

You may reopen this issue on the new fork. There is no need to continue this discussion here since this repo will be deleted soon. I've corrected the problem with the fork not having a issue tracker. (I honestly wasn't aware I could enable that for a fork. Github didn't do it automatically) so you should be able to open an issue there now. I'll close this issue for now as further discussion on this should occur there. Thanks for your input! :D

**Comment by ApacheThunder** [comment]
Lol nevermind on the above error. Removing references to the custom specs file in arm7's makefile and going back to the default one provided by libnds fixed it. (was going to do this anyways). So it compiles without error now. Will test that it still functions like it should. I assume it will. If I run into no problems this should be working on official libnds/devkitarm r46 this change will be pushed to the new fork soon™. :D

I did not have to change my maxmod related audio code either (So if there was updates to that it doesn't break with the update anyways. No surprise since the code is basically a stripped down copy and paste of the example code that came with it) so this was easier then I thought it would be. :P

**Comment by ApacheThunder** [comment]
Ok tested on hardware...bad news...it doesn't work. The bootsplash plays (and somehow it's running better then before? :P ) and it attempts to launch the cart, however it hangs. Not sure why. It doesn't hang on any of the debug color boxes in the debug display when I re-enabled that code which means the bootloader half of it isn't crashing. (note the only change I made to the SRL post compile is a small edit of the access control section in the DSi Extended Header. You have it compiling with the slot in ntr mode. However I do not enable that as with that bit set to 0 I'm able to init the card without power cycling the slot. Though I do provide a full card reset sequence via INI file config for users who still need to do that. I doubt this is the cause as I've had it setup like this in the previous commit and it worked fine then.

So either it's not loading the game binaries correctly, the game/cart in question crashes due to a problem in environment setup by NTR Launcher that is somehow borked by the new libnds version or some other issue I am not aware of. Maybe there's something it needs set in the arm7 SCFG_EXT section of the DSi Extended Header? Not sure. I can't use the values ndstool uses by default as that results in DSi touchscreen functions being active in NTR mode and breaks touchscreen on NTR games.

Until this is resolved I can't use this on libnds 1.6.2. I've committed the changes to the new fork for you to look at. Maybe I missed something.... 

I lack the means of properly debugging this so I won't be able to fix this alone. :(

EDIT: Tested using the default Arm7 SCFG_EXT values you are using for DSi Extended Header. This didn't alter it's behavior so I don't think that is the cause.

EDIT2: tested the SRL in a old build of hbmenu I use to toggle NTR mode with SD access still enabled. (to test if the mode switch that occurs in NTR Launcher is the breaking point). But this also doesn't change it's behavior besides it playing the DS version of the bootsplash instead of the DSi version. (cool that it appears to handle NTR mode with unlocked SCFG correctly though. Since the Bootsplash worked, it seems like your new detection code is working?)

So I don't think the mode switch is the cause. Maybe it breaks when the ram space is changed, but that has to occur else nothing would boot since NTR games can't handle that without extensive patches.

if that turns out to be the cause, I won't be able to use this on libnds 1.6.2 unless there's a major refractor where it can handle the ram space being changed back to NTR values. You said NitroHax worked....So I'm thinking this isn't the issue.

EDIT3: Tried the other branch, that one also hangs before the game/cart starts. So looks like the differences between the two branches are the cause of this problem.

**Comment by bakawun** [comment]
You should test nitrohax by building and trying to run in on your hardware to confirm that.

**Comment by ApacheThunder** [comment]
I suppose I could do that, but that would mean I would pretty much have to start over with it....Not really worth the effort. I could try and find the commit that might have broken it, but that too would be time consuming.

Keep in mind that WinterMute came to be about this issue when I was just about done updating this project. The only things I still wanted to do was fix the card init/read code (which is pretty old by this point) so that it can read/launch TWL carts correctly in NTR mode at the very least. Perhaps DSi Enhanced games could be run in TWL mode further down the road if that was ever fixed. But both of these things are out of scope of what I'm capable of doing myself.

I've already been wanting to get this to work with latest devkitarm r46 when I first saw the update for that drop. That will probably be the last major change I do to this project in the foreseeable future unless someone manages to send me a push request for improved boot animation code or the above mentioned fixes I've been wanting to do.

I'm just not motivated enough to completely redo everything just for the sake of getting this to work on latest devkitarm/libnds. If ultimately having to go back to square one just to this to work on the new kitchen sink update is needed then I just won't use r46. If r46 dropped earlier in my dev cycle for this perhaps I would have tackled it. After all patched TWL_FIRM has been around for how long now? Has to have been at least a year and half maybe. I don't recall. I suppose I can't expect the tool chains for stuff to keep up with that especially the older NTR/TWL stuff. :P

Anyways this project is sorta of in it's twilight years. If r47 comes around, don't expect me to bother with it unless someone sends me a push request for it that I don't have to bend over backwards to merge without a bunch of problems cropping up. I won't dig up old projects everytime devkitarm/libnds sees an update. I have to move on to other things eventually. :P

**Comment by bakawun** [comment]
What I meant to say is that you should confirm the claim that vanilla nitrohax works.

I suggest (also) compiling it yourself to make sure the reason for it not working isn't due to a difference between your developement environment and winter's

**Comment by ApacheThunder** [comment]
Ohh I see yeah I could take look to see if original NitroHax works or not. The build environment I setup for it is pretty basic. It was setup via devkitpro's installer which I had just about all the checkboxes enabled. It was compiled on Windows XP. The laptop is quite old by today's standards, but it gets the job done. I would have to perhaps install a few more things for CTR stuff to compile, but that's a whole other can of worms I didn't open because it wasn't necessary for the NTR/TWL stuff. :P

I could maybe check how original NitroHax gets along with it at least and confirm if that's the cause or not. That much I will do.

**Comment by bakawun** [comment]
Also the point you raise about not being able to keep this project updated indefinitey is the whole point behind this request. If these functions could be enabled inside nitrohax itself the larger community around that will help to keep it updated. If your changes to it are minimal, most of the time simply merging the changes from nitrohax would be enough to keep it updated.

Tl;dr : improving ease of maintenance is definately a goal here

**Comment by ApacheThunder** [comment]
Yeah I understand that. Problem is, it's a little too late now. I've just about moved on from this. :(

I'll fix the issues this has with r46...hopefully. But beyond that I'm probably done with it.

**Comment by ApacheThunder** [comment]
Ok got Nitrohax to work...after a lot of changes. Now that I think about it, unchanged it would have worked fine on original DS hardware. Chishm still hasn't updated it to work on 3DS however. So I had to re-add a few basic things just to get it to work like expected. First off I had it mode switch to NTR mode just after loading the cheat codes (as after this point SD access is no longer required. I used default NTR values to make sure it's a clean NTR environment). At first I didn't do the mode switch code at all. As a result the bootloader stage hung as expected as it wasn't adjusted to work in TWL mode.

That and I had to add in the slot reset code and add in the basic fifo wait code to make sure arm7/arm9 are on the same page on that regard. So it inits the card correctly? It immediately jumps to the correct code list for the game I inserted so at this stage it detected the cart correctly.

However once I hit start game, this is right around where the mode switch and reconfiguration of SCFG_EXT on arm7/arm9 occurs. Which I believe occurs properly (else the debug color boxes would never appear as they didn't when I hadn't added in the mode switch code)

But it ultimately still hangs on the red yellow yellow green box code (in that order). I dont' really know how to translate that to anything meaning ful. But it looks like the usual secure area CRC check fail or header check fail. Not sure which specifically, but that same error combo shows if you had booted it with no cart inserted.

Not sure why that's happening. Now that I think about it, I disabled the hault conditions for secure area check fail in NTR Launcher since some older flashcarts needed that disabled (it had no practical impact on retail games they would normally pass this check anyway)

So NTR Launcher zipped on past that error and tried to boot the game anyways which of coarse at this point fails.

I'm thinking it's not my build environment. Aside from the cart not launching, the cheat code loading/UI all operates correctly. I think the mode switch is for what ever reason causing the card loading stuff to fail. 

No idea why. Has to be something libnds did that is causing this.

At least I believe I know now where the hault is occuring. I guess WinterMute is correct that it works on latest libnds. But he wasn't specific. He must have tested it on original DS hardware where I suspect this still works. But I don't have an original DS to test this on, so can't fully confirm the integrity of my build environment.
